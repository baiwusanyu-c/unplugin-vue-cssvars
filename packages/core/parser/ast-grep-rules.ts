export const astGrepRules = {
  IDENT_NAME: {
    has: {
      kind: 'identifier',
      field: 'name',
    },
  },
  FN_CALL: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'call_expression',
      },
    },
  },
  ARROW_FN: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'arrow_function',
      },
    },
  },
  NOR_FN_VAR: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'function',
      },
    },
  },
  OBJ_VAR: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'object',
        field: 'value',
      },
    },
  },
  ARR_VAR: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'array',
        field: 'value',
      },
    },
  },
  NOR_FN: {
    has: {
      kind: 'function_declaration',
      has: {
        kind: 'identifier',
        field: 'name',
      },
    },
  },
  PROPS_DEFAULT_CALL: {
    has: {
      kind: 'call_expression',
      has: {
        kind: 'identifier',
        field: 'function',
      },
    },
  },
  PROPS_DEFAULT_ARG: {
    kind: 'property_identifier',
    inside: {
      kind: 'pair',
      inside: {
        kind: 'object',
        inside: {
          kind: 'arguments',
        },
      },
    },
  },
  PROPS_DEFAULT_VAL: {
    kind: 'object',
    inside: {
      kind: 'arguments',
    },
  },
  CONST_VAR: {
    any: [
      {
        pattern: 'const $VAR',
      },
    ],
  },
  LET_VAR: {
    any: [
      {
        pattern: 'let $VAR',
      },
    ],
  },
  CONST_REF_VAR: {
    any: [
      {
        pattern: 'const $VAR = ref($VAL)',
      },
    ],
  },
  CONST_REACTIVE_VAR: {
    any: [
      {
        pattern: 'const $VAR = reactive($VAL)',
      },
    ],
  },
  CONST_PROPS_VAR: {
    any: [
      {
        pattern: 'const $VAR = defineProps($VAL)',
      },
    ],
  },
  CONST_NOR_FN: {
    has:
      {
        kind: 'function_declaration',
      },
  },
  OBJ_PATTERN: {
    has:
      {
        kind: 'object_pattern',
      },
  },
  NEW_EXP: {
    has:
      {
        kind: 'new_expression',
        field: 'value'
      },
  },
  OBJ_PATTERN_VAL: {
    any:
      [
        {
          kind: 'identifier',
        },
        {
          kind: 'shorthand_property_identifier_pattern',
        },
      ],
  },
}

export const getRules = (name: string) => {
  return {
    rule: {
      matches: name,
    },
    utils: astGrepRules,
  }
}
