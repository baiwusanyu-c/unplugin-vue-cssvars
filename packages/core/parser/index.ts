export * from './parser-variable'
export * from './parser-import'
export * from './parser-vbindm'
export { parserCompiledSfc } from './parser-compiled-sfc'
