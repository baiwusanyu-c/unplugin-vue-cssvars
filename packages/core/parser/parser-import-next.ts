export enum ParserState {
  Initial,
  At,
  AtImport,
  AtUse,
  AtRequire,
  StringLiteral,
}

export interface ImportStatement {
  type: 'import' | 'use' | 'require'
  path: string
  start?: number
  end?: number
  suffix?: string
}
const delTransformSymbol = (content: string) => content.replace(/[\r\t\f\v\\]/g, '')
export function parseImportsNext(content: string): {
  imports: ImportStatement[]
  getCurState: () => ParserState
  getCurImport: () => undefined | ImportStatement
} {
  const imports: ImportStatement[] = []
  let currentImport: ImportStatement | undefined
  let state = ParserState.Initial
  let i = 0
  let AtPath = ''
  const source = delTransformSymbol(content)
  while (i < source.length) {
    const char = source[i]
    switch (state) {
      case ParserState.Initial:
        if (char === '@')
          state = ParserState.At
        break
      case ParserState.At:
        AtPath = AtPath + char
        if (AtPath === 'import') {
          AtPath = ''
          state = ParserState.AtImport
          currentImport = { type: 'import', path: '' }
          i++ // skip over "i" to next character
        } else if (AtPath === 'use') {
          AtPath = ''
          state = ParserState.AtUse
          currentImport = { type: 'use', path: '' }
          i++ // skip over "u" to next character
        } else if (AtPath === 'require') {
          AtPath = ''
          state = ParserState.AtRequire
          currentImport = { type: 'require', path: '' }
          i++ // skip over "u" to next character
        }
        // TODO ...
        if (AtPath.length >= 7 && AtPath !== 'require')
          state = ParserState.Initial

        break
      case ParserState.AtImport:
      case ParserState.AtUse:
      case ParserState.AtRequire:
        debugger
        // 当字符不是空格，且前一个是空格，进入取值
        if (char !== ' ' && source[i - 1] === ' ') {
          state = ParserState.StringLiteral
          currentImport!.start = i
          currentImport!.path += char
        } else if (char === '\n' || i === source.length - 1) {
          if (currentImport && currentImport.start !== undefined) {
            currentImport.end = i
            imports.push(currentImport)
            currentImport = undefined
          }
          state = ParserState.Initial
        }
        break
      case ParserState.StringLiteral:
        // 遇到引号，其起始位置也是，则是有引号状态下的取值结束
        if ((char === "'" || char === '"')
          && (currentImport!.start || currentImport!.start === 0)
          && (source[currentImport!.start] === char)) {
          if (currentImport!.type === 'import')
            state = ParserState.AtImport

          if (currentImport!.type === 'use')
            state = ParserState.AtUse

          if (currentImport!.type === 'require')
            state = ParserState.AtRequire

          currentImport!.path += char
        } else {
          // 取值
          currentImport!.path += char
        }

        break
    }
    i++
  }

  function getCurState() {
    return state
  }
  function getCurImport() {
    return currentImport
  }
  return {
    imports,
    getCurState,
    getCurImport,
  }
}
