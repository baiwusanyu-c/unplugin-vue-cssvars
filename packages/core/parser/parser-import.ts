const innerAtRule = 'media,extend,at-root,debug,warn,forward,mixin,include,function,error,keyframes,font-face,page,supports,namespace,return,if,else,for,while,each,content'
export enum ParserState {
  Initial,
  InlineComment,
  Comment,
  AtStart,
  AtEnd,
  AtImport,
  AtUse,
  AtRequire,
  QuotesStart,
  QuotesEnd,
  StringLiteral,
}

export interface ImportStatement {
  type: 'import' | 'use' | 'require'
  path: string
  start?: number
  end?: number
  suffix?: string
}

export function parseImports(content: string, helper?: Array<Function>): {
  imports: ImportStatement[]
  getCurState: () => ParserState
  getCurImport: () => undefined | ImportStatement
} {
  const imports: ImportStatement[] = []
  let currentImport: ImportStatement | undefined
  let state = ParserState.Initial
  let i = 0
  let AtPath = ''
  const source = content
  while (i < source.length) {
    const char = source[i]
    if (/[\r\t\f\v\\]/g.test(char)) {
      i++
      continue
    }

    switch (state) {
      case ParserState.Initial:
        if (char === '@')
          state = ParserState.AtStart
        if (char === '/' && source[i + 1] === '/')
          state = ParserState.InlineComment
        if (char === '/' && source[i + 1] === '*')
          state = ParserState.Comment
        break
      case ParserState.InlineComment:
        if (char === '\n')
          state = ParserState.Initial
        break
      case ParserState.Comment:
        if (char === '*' && source[i + 1] === '/')
          state = ParserState.Initial
        break
      case ParserState.AtStart:
        if (/[A-Za-z]$/.test(char))
          AtPath = AtPath + char
        else
          state = ParserState.AtEnd

        if (i === source.length - 1) {
          if (char === '"' || char === "'")
            throw new Error('syntax error: unmatched quotes')
          else
            walkContentEnd(i)
        }
        if (!(/[A-Za-z]$/.test(char))
          && char !== '\n'
          && char !== ' '
          && char !== '-')
          throw new Error('syntax error')

        break
      case ParserState.AtEnd:
        if (char !== '\n' && char !== ' ' && char !== '-') {
          if (char === '/')
            throw new Error('syntax error')

          if (AtPath === 'import') {
            AtPath = ''
            state = ParserState.AtImport
            currentImport = { type: 'import', path: '' }
            i--
          } else if (AtPath === 'use') {
            AtPath = ''
            state = ParserState.AtUse
            currentImport = { type: 'use', path: '' }
            i--
          } else if (AtPath === 'require') {
            AtPath = ''
            state = ParserState.AtRequire
            currentImport = { type: 'require', path: '' }
            i--
          } else {
            if (!innerAtRule.includes(AtPath))
              throw new Error('syntax error: unknown At Rule')

            AtPath = ''
            state = ParserState.Initial
          }
        }

        break
      case ParserState.AtImport:
      case ParserState.AtUse:
      case ParserState.AtRequire:
        // '@require test.css;@require test2.css'
        if (char === '@' && !(/[A-Za-z]$/.test(source[i - 1]))) {
          i--
          state = ParserState.Initial
          break
        }

        if (char === '/' && (source[i + 1] === '/' || source[i + 1] === '*')) {
          i--
          state = ParserState.Initial
          break
        }

        if (char === "'" || char === '"') {
          currentImport!.start = i
          state = ParserState.QuotesStart
          break
        }
        if (char !== '\n' && char !== ' ') {
          currentImport!.start = i
          currentImport!.path += char
          state = ParserState.StringLiteral
          break
        }
        break
      case ParserState.QuotesStart:
        if (char === "'" || char === '"') {
          currentImport!.end = i
          state = ParserState.QuotesEnd
          if (i === source.length - 1)
            walkContentEnd(i)
          break
        }
        if (i === source.length - 1)
          throw new Error('syntax error: unmatched quotes')

        currentImport!.path += char
        break
      case ParserState.QuotesEnd:
        if (i === source.length - 1 || char === ';' || char === '\n') {
          walkContentEnd(i)
        } else {
          i--
          state = ParserState.StringLiteral
        }
        break
      case ParserState.StringLiteral:
        if (char === ';' || char === '\n') {
          walkContentEnd(i)
          break
        }

        // '@require test.css@require test2.css'
        // '@import ./test1, ./test2'
        if (
          char !== '@'
          && (char === ' '
          || char === ','
          || char === '"'
          || char === "'")) {
          const curType = currentImport?.type
          walkContentEnd(i)
          if (curType === 'import') {
            state = ParserState.AtImport
            currentImport = { type: 'import', path: '' }
          } else if (curType === 'use') {
            state = ParserState.AtUse
            currentImport = { type: 'use', path: '' }
          } else if (curType === 'require') {
            state = ParserState.AtRequire
            currentImport = { type: 'require', path: '' }
          }

          if (char === "'" || char === '"') {
            currentImport!.start = i
            state = ParserState.QuotesStart
            if (i === source.length - 1 && (char === '"' || char === "'"))
              throw new Error('syntax error: unmatched quotes')

            break
          }

          break
        }

        currentImport!.path += char
        if (i === source.length - 1)
          walkContentEnd(i)

        break
    }
    i++
  }

  function walkContentEnd(index: number) {
    pushCurrentImport(index)
    state = ParserState.Initial
  }

  function pushCurrentImport(index: number) {
    if (currentImport && currentImport.start !== undefined) {
      currentImport.end = index
      if (helper) {
        helper.forEach((fn) => {
          currentImport = fn(currentImport)
        })
      }
      imports.push(currentImport)
      currentImport = undefined
    }
  }

  function getCurState() {
    return state
  }
  function getCurImport() {
    return currentImport
  }
  return {
    imports,
    getCurState,
    getCurImport,
  }
}
