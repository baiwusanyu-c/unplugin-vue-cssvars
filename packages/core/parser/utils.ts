// Enums:
/** This is derived from `@vue/compiler-core` */
import type { BindingTypes as VueBindingTypes } from '@vue/compiler-dom'
export const CSSVarsBindingTypes = {
  /**
   * returned from data()
   */
  DATA: 'data' as VueBindingTypes.DATA,
  /**
   * declared as a prop ✅
   */
  PROPS: 'props' as VueBindingTypes.PROPS,
  /**
   * a local alias of a `<script setup>` destructured prop.
   * the original is stored in __propsAliases of the bindingMetadata object.
   */
  PROPS_ALIASED: 'props-aliased' as VueBindingTypes.PROPS_ALIASED,
  /**
   * a let binding (may or may not be a ref) ✅
   */
  SETUP_LET: 'setup-let' as VueBindingTypes.SETUP_LET,
  /**
   * a const binding that can never be a ref.
   * these bindings don't need `unref()` calls when processed in inlined
   * template expressions. ✅
   */
  SETUP_CONST: 'setup-const' as VueBindingTypes.SETUP_CONST,
  /**
   * a const binding that does not need `unref()`, but may be mutated. ✅
   */
  SETUP_REACTIVE_CONST: 'setup-reactive-const' as VueBindingTypes.SETUP_REACTIVE_CONST,
  /**
   * a const binding that may be a ref. ✅
   */
  SETUP_MAYBE_REF: 'setup-maybe-ref' as VueBindingTypes.SETUP_MAYBE_REF,
  /**
   * bindings that are guaranteed to be refs ✅
   */
  SETUP_REF: 'setup-ref' as VueBindingTypes.SETUP_REF,
  /**
   * declared by other options, e.g. computed, inject
   */
  OPTIONS: 'options' as VueBindingTypes.OPTIONS,
  /**
   * a literal constant, e.g. 'foo', 1, true ✅
   */
  LITERAL_CONST: 'literal-const' as VueBindingTypes.LITERAL_CONST,
} as const
