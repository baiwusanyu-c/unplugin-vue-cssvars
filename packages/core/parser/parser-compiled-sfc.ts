import { parse as babelParse } from '@babel/parser'
import { walk } from 'estree-walker-ts'
import type {
  BlockStatement,
  CallExpression,
  Identifier,
  ImportSpecifier,
  ObjectExpression,
} from '@babel/types'
import type { Node } from 'estree-walker-ts'

export interface IParseSFCRes {
  setupBodyNode: BlockStatement
  hasCSSVars: boolean
  useCSSVarsNode: ObjectExpression
}

let isSetupEnter = false
let setupBodyNode = {} as BlockStatement
export function parseSetupBody(node: Node & { scopeIds?: Set<string> }) {
  if ((node as Identifier).type === 'Identifier' && (node as Identifier).name === 'setup') {
    isSetupEnter = true
    return
  }
  if (isSetupEnter && (node as BlockStatement).type === 'BlockStatement') {
    isSetupEnter = false
    setupBodyNode = node as BlockStatement
  }
}

let hasCSSVars = false
export function parseHasCSSVars(
  node: Node & { scopeIds?: Set<string> },
  parent: ImportSpecifier) {
  if ((node as Identifier).type === 'Identifier'
    && (node as Identifier).name === 'useCssVars'
    && parent
    && (((parent as ImportSpecifier).type) === 'ImportSpecifier'))
    hasCSSVars = true
}

let useCSSVarsNode = {} as ObjectExpression
let isUseCSSVarsEnter = false
export function parseUseCSSVars(
  node: Node & { scopeIds?: Set<string> },
  parent: CallExpression) {
  if ((node as Identifier).type === 'Identifier'
    && (node as Identifier).name === '_useCssVars'
    && parent
    && (((parent as CallExpression).type) === 'CallExpression'))
    isUseCSSVarsEnter = true

  if (isUseCSSVarsEnter && (node as ObjectExpression).type === 'ObjectExpression') {
    isUseCSSVarsEnter = false
    useCSSVarsNode = node as ObjectExpression
  }
}

export function parserCompiledSfc(code: string) {
  resetVar()
  const ast = babelParse(code, {
    sourceType: 'module',
    plugins: ['typescript'],
  });
  (walk as any)(ast, {
    enter(
      node: Node & { scopeIds?: Set<string> },
      parent: Node | undefined | ObjectExpression | ImportSpecifier | CallExpression) {
      parseSetupBody(node)
      parseHasCSSVars(node, parent as ImportSpecifier)
      parseUseCSSVars(node, parent as CallExpression)
    },
  })
  return {
    setupBodyNode,
    hasCSSVars,
    useCSSVarsNode,
  } as IParseSFCRes
}

export function resetVar() {
  isSetupEnter = false
  setupBodyNode = {} as BlockStatement
  isUseCSSVarsEnter = false
  useCSSVarsNode = {} as ObjectExpression
  hasCSSVars = false
}

// test only
export function getVar() {
  return {
    isSetupEnter,
    setupBodyNode,
    isUseCSSVarsEnter,
    useCSSVarsNode,
    hasCSSVars,
  }
}

// test only
export function setVar(k: string, v: any) {
  if (k === 'isSetupEnter')
    isSetupEnter = v

  if (k === 'setupBodyNode')
    setupBodyNode = v

  if (k === 'isUseCSSVarsEnter')
    isUseCSSVarsEnter = v

  if (k === 'useCSSVarsNode')
    useCSSVarsNode = v

  if (k === 'hasCSSVars')
    hasCSSVars = v
}
