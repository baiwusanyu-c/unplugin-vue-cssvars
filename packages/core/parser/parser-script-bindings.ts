import { ts } from '@ast-grep/napi'
import { CSSVarsBindingTypes } from './utils'
import { getRules } from './ast-grep-rules'
import type { SgNode } from '@ast-grep/napi'
import type { SFCDescriptor } from '@vue/compiler-sfc'
import type { BindingMetadata } from '@vue/compiler-dom'

export function analyzeScriptBindings(descriptor: SFCDescriptor): BindingMetadata {
  const scriptSetupContent = descriptor.scriptSetup?.content || ''
  const scriptContent = descriptor.script?.content || ''
  if (!scriptSetupContent && !scriptContent) return {}
  const bindings = {}
  const sgNodeScriptSetup = ts.parse(scriptSetupContent).root()
  walkSgNodeToGetBindings(sgNodeScriptSetup, bindings)
  const sgNodeScript = ts.parse(scriptContent).root()
  walkSgNodeToGetBindings(sgNodeScript, bindings)

  return bindings
}

function walkSgNodeToGetBindings(node: SgNode, bindings: BindingMetadata) {
  node.findAll(getRules('LET_VAR')).map((n) => {
    const key = n.getMatch('VAR')?.text() || ''
    bindings[key] = CSSVarsBindingTypes.SETUP_LET
  })

  let constVARs = node.findAll(getRules('CONST_VAR'))
  // function x(){}
  constVARs = constVARs.concat(node.findAll(getRules('CONST_NOR_FN')))
  constVARs.map((n) => {
    const key = n.getMatch('VAR')?.text() || n.find(getRules('IDENT_NAME'))?.text() || ''
    if (!key) return
    bindings[key] = CSSVarsBindingTypes.LITERAL_CONST
    // ref
    if (n.find(getRules('CONST_REF_VAR'))) {
      bindings[key] = CSSVarsBindingTypes.SETUP_REF
    }
    // reactive
    else if (n.find(getRules('CONST_REACTIVE_VAR'))) {
      bindings[key] = CSSVarsBindingTypes.SETUP_REACTIVE_CONST
    }
    // const a = b()、new xxx()
    else if (n.find(getRules('FN_CALL'))
      || n.find(getRules('NEW_EXP'))
      || (n.find(getRules('OBJ_PATTERN'))?.text().startsWith('{')
        && n.find(getRules('OBJ_PATTERN'))?.text().endsWith('}'))) {
      bindings[key] = CSSVarsBindingTypes.SETUP_MAYBE_REF
      //  解构赋值
      const deconstructVal = n.find(getRules('OBJ_PATTERN'))
      const deconstructKeyNode = deconstructVal?.findAll(getRules('OBJ_PATTERN_VAL'))
      deconstructKeyNode && deconstructKeyNode.forEach((nI) => {
        Reflect.deleteProperty(bindings, key)
        bindings[nI.text()] = CSSVarsBindingTypes.SETUP_MAYBE_REF
      })
    }
    // const a = () => {}
    // const a = function x(){}
    // function x(){}
    else if (
      n.find(getRules('ARROW_FN'))
      || n.find(getRules('NOR_FN'))
      || n.find(getRules('OBJ_VAR'))
      || n.find(getRules('ARR_VAR'))
      || n.kind() === 'function_declaration'
      || n.find(getRules('NOR_FN_VAR'))
    ) {
      if (n.kind() === 'function_declaration')
        bindings[n.find(getRules('IDENT_NAME'))?.text() || 'unk'] = CSSVarsBindingTypes.SETUP_CONST
      else
        bindings[key] = CSSVarsBindingTypes.SETUP_CONST
    }

    // const propss = withDefaults(defineProps<Props>(), {
    //   msg: 'hello',
    //   labels: () => ['one', 'two']
    // })
    // set 'msg'、'labels'
    if (n.find(getRules('PROPS_DEFAULT_CALL'))
      && n.find(getRules('PROPS_DEFAULT_CALL'))?.text().includes('withDefaults')) {
      const argObjNode = n.findAll(getRules('PROPS_DEFAULT_ARG'))
      argObjNode.forEach((nI) => {
        !bindings[nI.text()] && (bindings[nI.text()] = CSSVarsBindingTypes.PROPS)
      })
      bindings[key] = CSSVarsBindingTypes.SETUP_CONST
    }
    // e.g
    // const props = defineProps(
    //  {
    //    color: {
    //       type: String
    //     }
    //   })
    if (n.find(getRules('CONST_PROPS_VAR'))) {
      const argObjNode = n.findAll(getRules('PROPS_DEFAULT_ARG'))
      argObjNode.forEach((nI) => {
        !bindings[nI.text()] && (bindings[nI.text()] = CSSVarsBindingTypes.PROPS)
      })
      bindings[key] = CSSVarsBindingTypes.SETUP_REACTIVE_CONST
    }
  })
}
