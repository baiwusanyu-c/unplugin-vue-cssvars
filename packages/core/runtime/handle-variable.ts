import {babelParse, parse} from '@vue/compiler-sfc'
import { getVariable, matchVariable } from '../parser'
import { getVBindVariableListByPath } from './process-css'
import type { IVueCSSVarsCtx } from '../types'
import {analyzeScriptBindings} from "../parser/parser-script-bindings";

export function handleVBindVariable(
  code: string,
  id: string,
  ctx: IVueCSSVarsCtx,
) {
  const { descriptor } = parse(code)

  ctx.isScriptSetup = !!descriptor.scriptSetup
  // 分析 @import 的链路
  const {
    vbindVariableListByPath,
    injectCSSContent,
  } = getVBindVariableListByPath(
    descriptor,
    id,
    ctx.CSSFileModuleMap,
    ctx.isServer,
    ctx.userOptions.alias,
  )
  // 分析 sfc 内容中的变量
  const variableName = getVariable(descriptor)
  // 进行匹配得到最终的 cssvars 内容
  ctx.vbindVariableList.set(id, matchVariable(vbindVariableListByPath, variableName))
  // 解析 bindingsType
  ctx.isScriptSetup && (ctx.bindingsTypeMap[id] = analyzeScriptBindings(descriptor))
  return {
    descriptor,
    injectCSSContent,
  }
}
