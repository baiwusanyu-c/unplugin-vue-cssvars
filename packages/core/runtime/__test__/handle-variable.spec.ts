import { describe, expect, test } from 'vitest'
import { handleVBindVariable } from '../handle-variable'
import { CSSVarsBindingTypes } from '../../parser/utils'
import code from './sfc/mock-sfc.vue?raw'
const mockId = 'D:/project-github/unplugin-vue-cssvars/play/vite/src/views/app/App.vue'
const mockCode = code
const initMockCtx = () => {
  return {
    bindingsTypeMap: {
      [mockId]: {
        sassColor: CSSVarsBindingTypes.SETUP_REF,
        color: CSSVarsBindingTypes.SETUP_REF,
      },
    },
    CSSFileModuleMap: new Map(),
    vbindVariableList: new Map(),
    isScriptSetup: false,
    isServer: true,
    userOptions: {
      alias: {
        '@': 'D:\\project-github\\unplugin-vue-cssvars\\play\\vite\\src',
      },
    },
  }
}
const initMockCSSFileModuleMap = (mockContext: any) => {
  mockContext.CSSFileModuleMap.set(
    'D:/project-github/unplugin-vue-cssvars/play/vite/src/assets/css/foo.css',
    {
      vBindCode: ['sassColor', 'color'],
      content: '#foo{\n'
        + '  color: v-bind-m(sassColor);\n'
        + '  background: #ffebf8;\n'
        + '  width: 200px;\n'
        + '  height: 30px;\n'
        + '}\n'
        + 'p {\n'
        + '  color: v-bind-m(color);\n'
        + '}\n',
      lang: 'css',
      importer: new Set(),
    },
  )
}
describe('handle variable', () => {
  test('handleVBindVariable: basic', () => {
    const mockContext = initMockCtx()
    initMockCSSFileModuleMap(mockContext)
    handleVBindVariable(mockCode, mockId, mockContext as any)
    expect(mockContext.isScriptSetup).toBeTruthy()
    expect(mockContext.vbindVariableList.get(mockId)).toMatchObject([
      { value: 'sassColor', has: true, isRef: true },
      { value: 'color', has: true, isRef: true },
    ])
  })

  test('handleVBindVariable: server is false', () => {
    const mockContext = initMockCtx()
    mockContext.isServer = false
    initMockCSSFileModuleMap(mockContext)
    const res = handleVBindVariable(mockCode, mockId, mockContext as any)
    expect(mockContext.isScriptSetup).toBeTruthy()
    expect(mockContext.vbindVariableList.get(mockId)).toMatchObject([
      { value: 'sassColor', has: true, isRef: true },
      { value: 'color', has: true, isRef: true },
    ])
    expect(res.injectCSSContent).toMatchSnapshot()
  })
})
