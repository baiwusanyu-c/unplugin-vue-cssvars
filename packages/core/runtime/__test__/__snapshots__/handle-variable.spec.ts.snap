// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handle variable > handleVBindVariable: server is false 1`] = `
Set {
  {
    "content": "#foo{
  color: v-bind-m(sassColor);
  background: #ffebf8;
  width: 200px;
  height: 30px;
}
p {
  color: v-bind-m(color);
}
",
    "lang": "css",
    "styleTagIndex": 0,
  },
}
`;
